contracts/contracts/token/WOETH.sol

```diff
@@ -49,7 +90,84 @@ contract WOETH is ERC4626, Governable, Initializable {
     {
+        //@dev TODO: we could implement a feature where if anyone sends OETH direclty to
+        // the contract, that we can let the governor transfer the excess of the token.
         require(asset_ != address(asset()), "Cannot collect OETH");
         IERC20(asset_).safeTransfer(governor(), amount_);
     }
+
+    /**
+     * @dev This function converts requested OETH token amount to its underlying OETH
+     * credits value that is stored internally in OETH.sol and is required in order to
+     * be able to rebase.
+     *
+     * @param oethAmount Amount of OETH to be converted to OETH credits
+     * @return amount of OETH credits the OETH amount corresponds to
+     */
+    function _oethToCredits(uint256 oethAmount) internal returns (uint256) {
+        (, uint256 creditsPerTokenHighres, ) = OETH(asset())
+            .creditsBalanceOfHighres(address(this));
+
+        /**
+         * Multiplying OETH amount with the creditsPerTokenHighres is exactly the math that
+         * is internally being done in OETH:
+         */
+        // solhint-disable-next-line max-line-length
+        /** https://github.com/OriginProtocol/origin-dollar/blob/2314cccf2933f5c1f76a6549c1f5c9cc935b6f05/contracts/contracts/token/OUSD.sol#L242-L249
+         *
+         * This should make sure that the rounding will always be correct / mimic the rounding
+         * of OETH.
+         */
+        return oethAmount.mulTruncate(creditsPerTokenHighres);
+    }
+
+    /** @dev See {IERC4262-totalAssets} */
+    function totalAssets() public view virtual override returns (uint256) {
+        (, uint256 creditsPerTokenHighres, ) = OETH(asset())
+            .creditsBalanceOfHighres(address(this));
+
+        return (oethCreditsHighres).divPrecisely(creditsPerTokenHighres);
+    }
+
+    /** @dev See {IERC4262-deposit} */
+    function deposit(uint256 oethAmount, address receiver)
+        public
+        virtual
+        override
+        returns (uint256 woethAmount)
+    {
+        woethAmount = super.deposit(oethAmount, receiver);
+        oethCreditsHighres += _oethToCredits(oethAmount);
+    }
+
+    /** @dev See {IERC4262-mint} */
+    function mint(uint256 woethAmount, address receiver)
+        public
+        virtual
+        override
+        returns (uint256 oethAmount)
+    {
+        oethAmount = super.mint(woethAmount, receiver);
+        oethCreditsHighres += _oethToCredits(oethAmount);
+    }
+
+    /** @dev See {IERC4262-withdraw} */
+    function withdraw(
+        uint256 oethAmount,
+        address receiver,
+        address owner
+    ) public virtual override returns (uint256 woethAmount) {
+        woethAmount = super.withdraw(oethAmount, receiver, owner);
+        oethCreditsHighres -= _oethToCredits(oethAmount);
+    }
+
+    /** @dev See {IERC4262-redeem} */
+    function redeem(
+        uint256 woethAmount,
+        address receiver,
+        address owner
+    ) public virtual override returns (uint256 oethAmount) {
```

Comment:
This contract does not currently compile. This should be fixed.

Also, although this is not the actual compile error, I'm wondering if these `virtual`s in these methods are wrong, since I think we want these functions callable without needing to be overridden by a child class.

Reply:
Yes you are right these virtual keywords are not needed. We can add it in the future if need be: https://github.com/OriginProtocol/origin-dollar/pull/2106/commits/5ff5c5dfd0f3e81d386b6396894b954dfcf5f78b